java-checks/src/main/java/org/sonar/java/checks/unused/UnusedPrivateMethodCheck.java

```diff
@@ -20,41 +20,105 @@
 
+  private final List<MethodTree> unusedPrivateMethods = new ArrayList<>();
+  private final Set<String> unresolvedMethodNames = new HashSet<>();
+
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR);
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR, Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    super.scanFile(context);
+    reportUnusedPrivateMethods();
+    unusedPrivateMethods.clear();
+    unresolvedMethodNames.clear();
+  }
+
+  private void reportUnusedPrivateMethods() {
+    unusedPrivateMethods.stream()
+      .filter(methodTree -> !unresolvedMethodNames.contains(methodTree.simpleName().name()))
+      .forEach(methodTree -> {
+        IdentifierTree simpleName = methodTree.simpleName();
+        reportIssue(simpleName, String.format("Remove this unused private \"%s\" %s.", simpleName.name(), methodTree.is(Tree.Kind.CONSTRUCTOR) ? "constructor" : "method"));
+      });
   }
 
   @Override
   public void visitNode(Tree tree) {
     if(!hasSemantic()) {
       return;
     }
-    MethodTree node = (MethodTree) tree;
-    Symbol symbol = node.symbol();
-    if (node.modifiers().annotations().isEmpty() && symbol.isPrivate() && symbol.usages().isEmpty()) {
-      if (node.is(Tree.Kind.CONSTRUCTOR)) {
-        if (!node.parameters().isEmpty()) {
-          reportIssue(node.simpleName(), "Remove this unused private \"" + node.simpleName().name() + "\" constructor.");
+    if (tree.is(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR)) {
+      checkIfUnused((MethodTree) tree);
+    } else {
+      checkIfUnknown(tree);
+    }
+  }
+
+  private void checkIfUnknown(Tree tree) {
+    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
+      if (mit.symbol().isUnknown()) {
+        unresolvedMethodNames.add(MethodsHelper.methodName(mit).name());
+      }
+    } else {
+      NewClassTree nct = (NewClassTree) tree;
+      if (nct.constructorSymbol().isUnknown()) {
+        unresolvedMethodNames.add(constructorName(nct.identifier()));
+      }
+    }
+  }
+
+  private static String constructorName(TypeTree typeTree) {
+    if (typeTree.is(Tree.Kind.PARAMETERIZED_TYPE)) {
+      return constructorName(((ParameterizedTypeTree) typeTree).type());
+    } else if (typeTree.is(Tree.Kind.MEMBER_SELECT)) {
+      return ((MemberSelectExpressionTree) typeTree).identifier().name();
+    } else {
+      return ((IdentifierTree) typeTree).name();
+    }
+  }
+
+  private void checkIfUnused(MethodTree methodTree) {
+    Symbol symbol = methodTree.symbol();
+    if (methodTree.modifiers().annotations().isEmpty() && symbol.isPrivate() && symbol.usages().isEmpty()) {
+      if (methodTree.is(Tree.Kind.CONSTRUCTOR)) {
+        if (!methodTree.parameters().isEmpty()) {
```

Comment:
I would probably have a dedicated isConstructorWithNoParam method and have only one call to .add